pkg/dockerregistry/server/digestcache_test.go

```diff
@@ -0,0 +1,556 @@
+				},
+				{
+					repository: "pear",
+					evictOn:    now.Add(ttl5m),
+				},
+				{
+					repository: "apple",
+					evictOn:    now.Add(ttl8m),
+				},
+			},
+		},
+
+		{
+			name:  "add existing entry with lower ttl",
+			ttl:   ttl5m,
+			repos: []string{"orange"},
+			entries: []bucketEntry{
+				{
+					repository: "orange",
+					evictOn:    now.Add(ttl8m),
+				},
+				{
+					repository: "apple",
+					evictOn:    now.Add(ttl5m),
+				},
+				{
+					repository: "pear",
+					evictOn:    now.Add(ttl5m),
+				},
+			},
+			expectedEntries: []bucketEntry{
+				{
+					repository: "apple",
+					evictOn:    now.Add(ttl5m),
+				},
+				{
+					repository: "pear",
+					evictOn:    now.Add(ttl5m),
+				},
+				{
+					repository: "orange",
+					evictOn:    now.Add(ttl8m),
+				},
+			},
+		},
+
+		{
+			name:  "add new entry with eviction",
+			ttl:   ttl5m,
+			repos: []string{"banana"},
+			entries: []bucketEntry{
+				{
+					repository: "orange",
+					evictOn:    now.Add(ttl8m),
+				},
+				{
+					repository: "apple",
+					evictOn:    stale,
+				},
+				{
+					repository: "pear",
+					evictOn:    now.Add(ttl5m),
+				},
+			},
+			expectedEntries: []bucketEntry{
+				{
+					repository: "orange",
+					evictOn:    now.Add(ttl8m),
+				},
+				{
+					repository: "pear",
+					evictOn:    now.Add(ttl5m),
+				},
+				{
+					repository: "banana",
+					evictOn:    now.Add(ttl5m),
+				},
+			},
+		},
+
+		{
+			name:  "all stale",
+			ttl:   ttl5m,
+			repos: []string{"banana"},
+			entries: []bucketEntry{
+				{
+					repository: "orange",
+					evictOn:    stale,
+				},
+				{
+					repository: "apple",
+					evictOn:    stale,
+				},
+				{
+					repository: "pear",
+					evictOn:    stale,
+				},
+			},
+			expectedEntries: []bucketEntry{
+				{
+					repository: "banana",
+					evictOn:    now.Add(ttl5m),
+				},
+			},
+		},
+
+		{
+			name:  "add multiple entries with middle ttl",
+			ttl:   ttl5m,
+			repos: []string{"apple", "banana", "peach", "orange"},
+			entries: []bucketEntry{
+				{
+					repository: "melon",
+					evictOn:    stale,
+				},
+				{
+					repository: "orange",
+					evictOn:    now.Add(ttl8m),
+				},
+				{
+					repository: "apple",
+					evictOn:    now.Add(ttl5m),
+				},
+				{
+					repository: "pear",
+					evictOn:    now.Add(ttl1m),
+				},
+				{
+					repository: "plum",
+					evictOn:    stale,
+				},
+			},
+			expectedEntries: []bucketEntry{
+				{
+					repository: "pear",
+					evictOn:    now.Add(ttl1m),
+				},
+				{
+					repository: "apple",
+					evictOn:    now.Add(ttl5m),
+				},
+				{
+					repository: "banana",
+					evictOn:    now.Add(ttl5m),
+				},
+				{
+					repository: "peach",
+					evictOn:    now.Add(ttl5m),
+				},
+				{
+					repository: "orange",
+					evictOn:    now.Add(ttl8m),
+				},
+			},
+		},
+
+		{
+			name:    "over bucket size",
+			ttl:     ttl1m,
+			repos:   []string{"new1", generated[2].repository, "new2", generated[4].repository},
```

Comment:
add a test where we're more than one over the bucketsize and make sure we get trimmed down to exactly the bucket size